server/src/main.rs

```diff
@@ -0,0 +1,156 @@
+    static ref X_ORIGINAL_URI: HeaderName = HeaderName::from_lowercase(b"x-original-uri").unwrap();
+    static ref HP_PUBLIC_KEY: PublicKey = read_hp_pubkey();
+}
+
+// Create response based on the request parameters
+fn create_response(req: Request<Body>) -> impl Future<Item = Response<Body>, Error = hyper::Error> {
+    let (parts, body) = req.into_parts();
+
+    match parts.uri.path() {
+        "/" => {
+            let entire_body = body.concat2();
+            let res = entire_body.map( |body| {
+                // Extract X-Original-URI header value, panic for no header
+                let req_uri = match parts.headers.get(&*X_ORIGINAL_URI) {
+                    Some(s) => s.to_str().unwrap(),
+                    None => panic!("Request does not contain \"X-Original-URI\" header."),
+                };
+                let body_string = String::from_utf8(body.to_vec()).expect("Found invalid UTF-8");
+                let payload = create_payload(parts.method.to_string(), req_uri.to_string(), body_string);
+                let is_verified = verify_request(payload, parts.headers, &HP_PUBLIC_KEY);
+                respond_success(is_verified)
+            });
+
+            Either::A(res)
+        }
+        _ => {
+            let res = future::ok(respond_success(false));
+            Either::B(res)
+        }
+    }
+}
+
+fn create_payload (method: String, uri: String, body_string: String) -> String {
+    let d = json!({
+        "method": method.to_lowercase(), // make sure verb is to lowercase
+        "uri": uri,
+        "body": body_string
+    }); 
+
+    // Serialize it to a JSON string.
+    serde_json::to_string(&d).unwrap()
+}
+
+fn verify_request(payload: String, headers: HeaderMap<HeaderValue>, public_key: &PublicKey) -> bool {
+    if let Some(signature_base64) = headers.get(&*X_HPOS_ADMIN_SIGNATURE) {
+        if let Ok(signature_vec) = decode_config(&signature_base64, base64::STANDARD_NO_PAD) {
+            if let Ok(signature_bytes) = Signature::from_bytes(&signature_vec) {
+                if public_key.verify(&payload.as_bytes(), &signature_bytes).is_ok() {
+                    return true
+                }
+            }
+        }
+    }
+    return false
+}
+
+fn respond_success (is_verified: bool) -> hyper::Response<Body> {
+    // construct response based on verification status
+    match is_verified {
+        true => {
+            Response::builder()
+                .status(StatusCode::OK)
+                .body(Body::empty())
+                .unwrap()
+        },
+        _ => {
+            Response::builder()
+                .status(StatusCode::UNAUTHORIZED)
+                .body(Body::empty())
+                .unwrap()
+        }
+    }
+}
+
+fn read_hp_pubkey() -> PublicKey {
+    let hpos_state_path = env::var("HPOS_STATE_PATH").expect("HPOS_STATE_PATH environmental variable is not present");
+
+    println!("Reading HP Admin Public Key from {}.", hpos_state_path);
```

Comment:
At the very least:

```suggestion
    eprintln!("Reading HP Admin Public Key from {}.", hpos_state_path);
```

Better yet, this should use `log` crate.

Reply:
True, you want this to be an error message

Reply:
Actually no - this conversation only makes sense in the context: what do we log, where do we log and what for do we log.